### PR TITLE
Fix: Function by method with the same name as a method won't crash resolver

### LINKED ISSUE #2019: function-by-method crash
Dafny correctly notices a duplicate method name, but then crashes. This is unlike regular functions and methods.

```
method f(x: seq<int>) returns (res: seq<int>)
  ensures x == res
{
  var y := [];
  for i := 0 to |x|
    invariant |y| == i
    invariant x[..i] == y[..i]
  {
    y := y + [x[i]];
  }
  return y;
}

function f(x: seq<int>): seq<int> {
  x
} by method {
  var y := [];
  for i := 0 to |x|
    invariant |y| == i
    invariant x[..i] == y[..i]
  {
    y := y + [x[i]];
  }
  return y;
}
```

```
$ dafny /compile:0 /tmp/foo.dfy
/tmp/foo.dfy(14,9): Error: Duplicate member name: f
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Dafny.Resolver.ResolveMethod(Method m) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/Dafny/Resolver.cs:line 10254
   at Microsoft.Dafny.Resolver.ResolveFunction(Function f) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/Dafny/Resolver.cs:line 10164
   at Microsoft.Dafny.Resolver.ResolveClassMemberBodies(TopLevelDeclWithMembers cl) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/Dafny/Resolver.cs:line 9559
   at Microsoft.Dafny.Resolver.ResolveTopLevelDecls_Core(List`1 declarations, Graph`1 datatypeDependencies, Graph`1 codatatypeDependencies, Boolean isAnExport) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/Dafny/Resolver.cs:line 2811
   at Microsoft.Dafny.Resolver.ResolveModuleDefinition(ModuleDefinition m, ModuleSignature sig, Boolean isAnExport) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/Dafny/Resolver.cs:line 1024
   at Microsoft.Dafny.Resolver.ResolveProgram(Program prog) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/Dafny/Resolver.cs:line 497
   at Microsoft.Dafny.Main.Resolve(Program program, ErrorReporter reporter) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/Dafny/DafnyMain.cs:line 161
   at Microsoft.Dafny.DafnyDriver.ProcessFiles(ExecutionEngineOptions options, IList`1 dafnyFiles, ReadOnlyCollection`1 otherFileNames, ErrorReporter reporter, Boolean lookForSnapshots, String programId) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/DafnyDriver/DafnyDriver.cs:line 290
   at Microsoft.Dafny.DafnyDriver.ThreadMain(String[] args) in /Users/salkeldr/Documents/GitHub/release-dafny/Source/DafnyDriver/DafnyDriver.cs:line 57
   at Microsoft.Dafny.DafnyDriver.<>c__DisplayClass1_0.<Main>b__0() in /Users/salkeldr/Documents/GitHub/release-dafny/Source/DafnyDriver/DafnyDriver.cs:line 37
   at System.Threading.Thread.StartCallback()
[3]    36446 abort
```

### PR DESCRIPTION
**Pull request by MikaelMayer**
This PR fixes #2019
I added the corresponding test.
The story was that, if a function by method and a method had the same name, the function by method's resolved method was null. However, there were two places in the resolver, before it gives up, that were asserting that the function by method's method was not null, so I added a conditional instead and ensured that the number of reported errors was greater than zero.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/Resolver.cs [hunk 1]
```diff
@@ -10289,8 +10295,14 @@ public CheckDividedConstructorInit_Visitor(Resolver resolver)
...
           var method = f.ByMethodDecl;
-          Contract.Assert(method != null); // this should have been filled in by now
-          ResolveMethod(method);
+          if (method != null) {
+            ResolveMethod(method);
+          } else {
+            // method should have been filled in by now,
+            // unless there was a function by method and a method of the same name
+            // but then this error must have been reported.
+            Contract.Assert(reporter.ErrorCount >= 1);
```
Comment (alex-chew) [inline Source/DafnyCore/Resolver.cs:10304]: nit: Why `>= 1` here but `> 0` in the other case?

**Review by alex-chew — DISMISSED** [review]
LGTM, added one non-blocking comment

File: Source/DafnyCore/Resolver.cs (hunk 1 above)
Comment (MikaelMayer) [inline Source/DafnyCore/Resolver.cs:10304]:
```suggestion
            Contract.Assert(reporter.ErrorCount > 0);
```

Comment (MikaelMayer) [reply Source/DafnyCore/Resolver.cs:10304]: > nit: Why >= 1 here but > 0 in the other case?

i did not do them at the same time :-) will fix it for consistency.

**Review by alex-chew — APPROVED** [review]

**Pull request merged**